src/server/replication_messages.rs

```diff
@@ -115,52 +117,302 @@ impl ReplicationMessages {
+    /// See also [`Self::start_entity_data`].
+    pub(super) fn write_replication_id(
+        &mut self,
+        replication_id: ReplicationId,
+    ) -> bincode::Result<()> {
+        if self.entity_data_size == 0 {
+            self.write_data_entity()?;
+        }
+
+        let previous_pos = self.cursor.position();
+        DefaultOptions::new().serialize_into(&mut self.cursor, &replication_id)?;
+
+        let id_size = self.cursor.position() - previous_pos;
+        self.entity_data_size = self
+            .entity_data_size
+            .checked_add(id_size as u16)
+            .ok_or(bincode::ErrorKind::SizeLimit)?;
+
+        Ok(())
+    }
+
+    /// Removes entity data elements from update message and copies it.
+    ///
+    /// Ends entity data for the update message.
+    /// See also [`Self::start_entity_data`] and [`Self::end_entity_data`].
+    pub(super) fn take_entity_data(
+        &mut self,
+        update_message: &mut UpdateMessage,
+    ) -> bincode::Result<()> {
+        if update_message.entity_data_size != 0 {
+            let slice = update_message.as_slice();
+            let offset = update_message.entity_data_size_pos as usize
+                + mem::size_of_val(&update_message.entity_data_size);
+            self.cursor.write_all(&slice[offset..]).unwrap();
+
+            self.entity_data_size = self
+                .entity_data_size
+                .checked_add(update_message.entity_data_size)
+                .ok_or(bincode::ErrorKind::SizeLimit)?;
+            update_message.entity_data_size = 0;
+        }
+
+        update_message
+            .cursor
+            .set_position(update_message.entity_data_pos);
+
+        Ok(())
+    }
+
+    /// Returns the serialized data, excluding trailing empty arrays, as a byte array.
+    fn as_slice(&self) -> &[u8] {
+        let slice = self.cursor.get_ref();
+        let position = self.cursor.position() as usize;
+        let extra_len = self.trailing_empty_arrays * mem::size_of_val(&self.array_len);
+        &slice[..position - extra_len]
+    }
+
+    /// Sends the message, excluding trailing empty arrays, to the specified client.
     ///
     /// Does nothing if there is no data to send.
-    fn send(&mut self, server: &mut RenetServer, client_id: ClientId) {
-        if !self.is_sendable() {
+    fn send(
+        &self,
+        server: &mut RenetServer,
+        replicon_tick: RepliconTick,
+        client_id: ClientId,
+    ) -> bincode::Result<()> {
+        debug_assert_eq!(self.array_len, 0);
+        debug_assert_eq!(self.entity_data_size, 0);
+
+        let slice = self.as_slice();
+        if slice.is_empty() {
             trace!("no init data to send for client {client_id}");
-            return;
+            return Ok(());
         }
 
-        self.buffer.trim_empty_arrays();
+        let mut header = [0; mem::size_of::<RepliconTick>()];
+        bincode::serialize_into(&mut header[..], &replicon_tick)?;
```

Comment:
I was going to say that both iterators are `ExactSizeIterator`, but looks like [`Chain` is not](https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html#when-shouldnt-an-adapter-be-exactsizeiterator).
So I agree with the suggestion. But I used `concat` since it's a bit shorter and more explicit about the header and the rest of the message, especially for `UpdateMessage`.